### ISSUE TITLE
Using `NextAddr` always returns new address

### ISSUE THREAD

**Issue by kaganovych** [lightningnetwork/lnd#6118]
### Background

There're 2 ways to generate new address:
From Lnrpc: https://api.lightning.community/#newaddress
And from walletkit: https://api.lightning.community/#nextaddr

As I understand the difference is that the latter returns the same address until it's used. But both of them return a new address each time I call the request. Is it an intended behavior or a bug? 

### Your environment

* version of `lnd` - v0.14.1-beta
* which operating system (`uname -a` on *Nix) - Android/iOS

### Steps to reproduce

I'm developing the Android app, so I use either `LightningOuterClass.NewAddressRequest` or `Walletkit.AddrRequest`.

### Expected behaviour

`Walletkit.AddrRequest` should return the same address until it's used.

### Actual behaviour

`Walletkit.AddrRequest` always returns the new generated address.

**Comment by alexbosworth** [comment]
When you request a format of the address you can request the used or unused format

**Comment by kaganovych** [comment]
@alexbosworth could you please specify? In `NextAddr` I can configure `type` and `change` value. Are you talking about type?

**Comment by alexbosworth** [comment]
If you call the `NewAddress` RPC then you will provide a type of address you want:

```
message NewAddressRequest {
    // The type of address to generate.
    AddressType type = 1;

    /*
    The name of the account to generate a new address for. If empty, the
    default wallet account is used.
    */
    string account = 2;
}
```

The type is defined as an enum that includes unused type to get a new address on invocation

```
enum AddressType {
    WITNESS_PUBKEY_HASH = 0;
    NESTED_PUBKEY_HASH = 1;
    UNUSED_WITNESS_PUBKEY_HASH = 2;
    UNUSED_NESTED_PUBKEY_HASH = 3;
}
```

**Comment by kaganovych** [comment]
Sorry, but from 0.14.1-beta I have next:

```
enum AddressType {
    UNKNOWN = 0;
    WITNESS_PUBKEY_HASH = 1;
    NESTED_WITNESS_PUBKEY_HASH = 2;
    HYBRID_NESTED_WITNESS_PUBKEY_HASH = 3;
}
```

**Comment by alexbosworth** [comment]
That's in walletkit, did you want a new address from the regular new address API?

**Comment by kaganovych** [comment]
@alexbosworth I'll try that tomorrow and will return with a result, thanks.

**Comment by kaganovych** [comment]
@alexbosworth yes, the flag with the `UNUSED` prefix works. Thanks a lot!

**Comment by schjonhaug** [comment]
@alexbosworth 

> That's in walletkit, did you want a new address from the regular new address API?

What is the difference between getting an address from the WalletKit vs. the Lightning service? The docs are not clear, and they suggest using NextAddr if we want an unused one:

Service Lightning
[NewAddress](https://api.lightning.community/#newaddress):

> NewAddress creates a new address under control of the local wallet.

Service WalletKit
[NextAddr](https://api.lightning.community/#nextaddr)

> NextAddr returns the next unused address within the wallet.

Not sure what is meant by ’local wallet’ and ‘within the wallet‘ here. We are looking for an unused address to store sats on-chain.

**Comment by alexbosworth** [comment]
WalletKit is an experimental API so I would use the Lightning service

**Comment by schjonhaug** [comment]
@alexbosworth Ok, thanks. Could the WalletKit API be marked as experimental to avoid confusion?

**Comment by alexbosworth** [comment]
Maybe a new issue to clarify documentation

**Comment by guggero** [comment]
> @alexbosworth Ok, thanks. Could the WalletKit API be marked as experimental to avoid confusion?

It is marked as experimental: https://api.lightning.community/#experimental-services

**Comment by Roasbeef** [comment]
Correct that they both return a fresh address. So it's a documentation issue.

**Comment by schjonhaug** [comment]
These are the issues I’ve discovered in the docs:

## Non-working link

<img width="619" alt="Skjermbilde 2022-01-04 kl  21 16 05" src="https://user-images.githubusercontent.com/84225/148118469-e00f972e-4aec-47fa-a672-5161c63cb30a.png">


The link to `AddressType` at https://api.lightning.community/#newaddress points to the non-existing URL https://api.lightning.community/#lnrpc-addresstype. The correct URL should be https://api.lightning.community/#addresstype


## Unusual acronyms

The shortened acronyms of these address types are a bit unusual, and would be less ambigious if including the missing ’public’ part, i.e. `p2wpkh`: Pay to witness public key hash.

```
/*
`AddressType` has to be one of:
- `p2wkh`: Pay to witness key hash (`WITNESS_PUBKEY_HASH` = 0)
- `np2wkh`: Pay to nested witness key hash (`NESTED_PUBKEY_HASH` = 1)
*/

```

## Incorrect number of options for the `AddressType` enum

The docs above on `AddressType` claims to only have two options, but in reality the [enum has four](https://github.com/lightningnetwork/lnd/blob/master/lnrpc/lightning.proto#L1105).  Since this is currently the only way to get the (same) unused address each time, I think this needs to be highlighted.

```
enum AddressType {
    WITNESS_PUBKEY_HASH = 0;
    NESTED_PUBKEY_HASH = 1;
    UNUSED_WITNESS_PUBKEY_HASH = 2;
    UNUSED_NESTED_PUBKEY_HASH = 3;
}
```


## Lack of clarity in how to get an unused address

Ignoring the fact that WalletKit is experimental, the docs give the impression that `NextAddr` is the way to get “the next unused address“

From https://api.lightning.community/#nextaddr

> NextAddr returns the next unused address within the wallet.

But it’s unclear if we get a new unused address each time we call the function, or the same unused one each time. The same can be said about the aforementioned `NewAddress`: it’s unclear if we get the same unused address each time, or the next one.

**Comment by schjonhaug** [comment]
Related to the address type issue mentioned above, I have discovered a discrepancy in the way the RPC and REST API works with regards to newAddress.

I’m using LND using the official Docker image. Running `docker exec -it lnd-regtest lncli --no-macaroons newaddress` I cannot see any way to get an unused address:

```
NAME:
   lncli newaddress - Generates a new address.

USAGE:
   lncli newaddress [command options] address-type

CATEGORY:
   Wallet

DESCRIPTION:

  Generate a wallet new address. Address-types has to be one of:
      - p2wkh:  Pay to witness key hash
      - np2wkh: Pay to nested witness key hash

OPTIONS:
   --account value  (optional) the name of the account to generate a new address for
```

However, using curl (inside the container), I can use the `UNUSED_WITNESS_PUBKEY_HASH` and get the same unused address every time:

```
/ # curl -X GET --cacert /root/.lnd/tls.cert https://localhost:8080/v1/newaddress?type=UNUSED_WITNESS_PUBKEY_HASH
{"address":"bcrt1qhmzjfta6ky92eec7ukkyqy59nqqxcq63l0gwhv"}/ #

/ # curl -X GET --cacert /root/.lnd/tls.cert https://localhost:8080/v1/newaddress?type=UNUSED_WITNESS_PUBKEY_HASH
{"address":"bcrt1qhmzjfta6ky92eec7ukkyqy59nqqxcq63l0gwhv"}/ #

/ # curl -X GET --cacert /root/.lnd/tls.cert https://localhost:8080/v1/newaddress?type=UNUSED_WITNESS_PUBKEY_HASH
{"address":"bcrt1qhmzjfta6ky92eec7ukkyqy59nqqxcq63l0gwhv"}/ #
```

**Comment by guggero** [comment]
That's probably just because `lncli` wasn't updated after those "unused" types were added. We could add a simple `--unused` flag to it, that makes it use those types.

Make sure that you don't use this feature (unused addresses) in combination with Neutrino. Because Neutrino doesn't see transactions in the mempool, an address would only be recognized as "used" after the transaction confirms. Since that can take a while, the user might still use that address again and involuntarily do address-reuse.

**Comment by schjonhaug** [comment]
Yes, feature parity would be nice for both `lncli` and the REST API.

Regarding the lack of mempool using Neutrino: Could LND cache the sent transactions locally until they show up in the blockchain? They do get transaction ids and can be viewed in a block explorer, right? From a user’s perspective, I would reckon it be frightening sending off some coins without them immediately showing up in the transactions list as unconfirmed/pending.

**Comment by gabbyprecious** [comment]
@alexbosworth is this currently being worked on? Or can I be assigned to it?

**Comment by guggero** [comment]
You can just go ahead and create a PR.
I think there are three main tasks:
 - Update `lncli` and add a `--unused` flag that switches from `AddressType_WITNESS_PUBKEY_HASH` to `AddressType_UNUSED_WITNESS_PUBKEY_HASH` for example
 - Update the comment of the `AddressType` enum in `lightning.proto` to include the "unused" types
 - Fix https://github.com/lightninglabs/lightning-api to generate correct links to the enum types (the API docs for `NewAddress` should link to https://api.lightning.community/#lnrpcaddresstype instead of https://api.lightning.community/#lnrpc-addresstype)

**Comment by gabbyprecious** [comment]
Hi @guggero I sent a draft pr 
https://github.com/lightningnetwork/lnd/pull/6375
tasks
- in pr. 
- saw that this has been done by you. 
- would like more info on how this can be done, went through the repo and didn't see anything.
